src/hazelcore/UHExp.re

```diff
@@ -308,10 +308,130 @@ let text_operand =
+       */
+      go_affix(
+        [Skel.Placeholder(lex_addr), ...skels],
+        op_stack,
+        affix, /* Tail of seq without first operand. */
+        lex_addr + 1 /* Increment lexical address of operand in seq. */
+      )
+    };
+  }
+  and go_affix =
+      (
+        skels: list(skel),
+        op_stack: list(Operators_Exp.t),
+        affix: affix,
+        lex_addr: int,
+      )
+      : skel => {
+    switch (affix) {
+    | A(op, seq) =>
+      /**
+       * If the next token is an operator, pop operators from the operator stack,
+       * then push this operator to the operator stack.
+       */
+      let should_mv = op' =>
+        /**
+         * Continue popping operators while the precedence of the top of the operator
+         * stack has greater precedence than the current operator.
+         *
+         * Either the operator on the top of the stack is left associative and
+         * has greater or equal precedence to the current operator, or the
+         * operator on the top of the stack is right associative and has strictly
+         * greater precedence to the current operator.
+         */
+        {
+          switch (Operators_Exp.associativity(op')) {
+          | Left =>
+            Operators_Exp.precedence(op) <= Operators_Exp.precedence(op')
+          | Right =>
+            Operators_Exp.precedence(op) < Operators_Exp.precedence(op')
+          };
+        };
+
+      let (skels', op_stack') = build_ast_while(skels, op_stack, should_mv);
+
+      /* Push this operator to the operator stack. */
+      go_seq(skels', [op, ...op_stack'], seq, lex_addr);
+    | E =>
+      /**
+       * Once the input seq is empty, continuously pop
+       * operators in the stack and build up the output skel.
+       */
+      let (skels', _) = build_ast_while(skels, op_stack, _ => true);
+      switch (skels') {
+      | [final_skel] => final_skel // In this case,
+      | _ => Skel.Placeholder(-1) // This case will never be reached
+      };
+    };
+  }
+  and build_ast_while =
+      (skels: list(skel), op_stack: list(Operators_Exp.t), should_mv)
+      : (list(skel), list(Operators_Exp.t)) => {
+    /* Move operators from the operator stack to the output skel list while... */
+    switch (op_stack) {
+    | [] => (skels, op_stack) /* (1) The operator stack is not empty. */
+    | [op, ...op_stack'] =>
+      if (should_mv(op)) {
+        /**  (2) See defn of should_mv in go_affix.
+         *       Note - This impl supports only binary operators.
+         *
+         * Example -
+         *
+         * skels:
+         * [Placeholder(0) Placeholder(1)]
+         *
+         * op_stack:
+         * [Plus]
+         *
+         * -->
+         *
+         * skels:
+         * [BinOp(_, Plus, Placeholder(1), Placeholder(0))]
+         *
+         * op_stack:
+         * []
+         *
+         */
+        (
+          switch (skels) {
+          | [rand_2, rand_1, ...skels'] =>
```

Comment:
By `rand`, I assume you mean `operand`? I'd call these subskels, we use operand to mean specifically the things you find at `Placeholder` nodes.

Reply:
Sure thing.